File: python/database_operations.py
import sqlite3
import pandas as pd

def query_database(db_path: str, query: str) -> pd.DataFrame:
    """
    Run a query on an sqlite database and return the results in a dataframe
    
    Args:
        db_path: Path to the SQLite database file
        query: SQL query to execute
        
    Returns:
        pandas.DataFrame containing the query results
    """
    with sqlite3.connect(db_path) as conn:
        return pd.read_sql_query(query, conn)


def execute_non_query_statement(db_path: str, statement: str, params: tuple = ()) -> int:
    """
    Execute a non-query SQL statement (INSERT, UPDATE, DELETE)
    
    Args:
        db_path: Path to the SQLite database file
        statement: SQL statement to execute
        params: Parameters for the statement
        
    Returns:
        Number of affected rows
    """
    with sqlite3.connect(db_path) as conn:
        cursor = conn.cursor()
        cursor.execute(statement, params)
        conn.commit()
        return cursor.rowcount


def execute_scalar(db_path: str, query: str, params: tuple = ()) -> any:
    """
    Execute a query and return the first column of the first row
    
    Args:
        db_path: Path to the SQLite database file
        query: SQL query to execute
        params: Parameters for the query
        
    Returns:
        The value of the first column of the first row, or None
    """
    with sqlite3.connect(db_path) as conn:
        cursor = conn.cursor()
        cursor.execute(query, params)
        result = cursor.fetchone()
        return result[0] if result else None


def get_table_names(db_path: str) -> list[str]:
    """
    Get a list of all tables in the database
    
    Args:
        db_path: Path to the SQLite database file
        
    Returns:
        List of table names
    """
    query = "SELECT name FROM sqlite_master WHERE type='table'"
    with sqlite3.connect(db_path) as conn:
        cursor = conn.cursor()
        cursor.execute(query)
        return [row[0] for row in cursor.fetchall()]


def table_exists(db_path: str, table_name: str) -> bool:
    """
    Check if a table exists in the database
    
    Args:
        db_path: Path to the SQLite database file
        table_name: Name of the table to check
        
    Returns:
        True if table exists, False otherwise
    """
    return table_name in get_table_names(db_path)


def get_row_count(db_path: str, table_or_query: str) -> int:
    """
    Get the row count of a table or query
    
    Args:
        db_path: Path to the SQLite database file
        table_or_query: Name of the table or a SQL SELECT query
        
    Returns:
        Number of rows
    """
    if table_or_query.strip().upper().startswith("SELECT"):
        query = f"SELECT COUNT(1) FROM ({table_or_query})"
    else:
        query = f"SELECT COUNT(1) FROM {table_or_query}"
    
    return execute_scalar(db_path, query)

File: dotNet/DatabaseOperations.cs
using System;
using System.Collections.Generic;
using System.Data;
using Microsoft.Data.SqlClient;

namespace TestAutomation.Utils
{
    /// <summary>
    /// SQL Server database operations utility functions for test automation
    /// </summary>
    public static class DatabaseOperations
    {
        /// <summary>
        /// Run a query on a SQL Server database and return the results
        /// </summary>
        /// <param name="connectionString">SQL Server connection string</param>
        /// <param name="query">SQL query to execute</param>
        /// <returns>DataTable containing the query results</returns>
        public static DataTable QueryDatabase(string connectionString, string query)
        {
            try
            {
                using (SqlConnection conn = new SqlConnection(connectionString))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(query, conn))
                    {
                        using (SqlDataAdapter adapter = new SqlDataAdapter(cmd))
                        {
                            DataTable dt = new DataTable();
                            adapter.Fill(dt);
                            return dt;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error executing query: {ex.Message}");
                throw;
            }
            finally
            {
                // Additional cleanup if needed
            }
        }

        /// <summary>
        /// Execute a non-query SQL statement (INSERT, UPDATE, DELETE)
        /// </summary>
        /// <param name="connectionString">SQL Server connection string</param>
        /// <param name="statement">SQL statement to execute</param>
        /// <param name="params">Parameters for the statement</param>
        /// <returns>Number of affected rows</returns>
        public static int ExecuteNonQuery(string connectionString, string statement, params SqlParameter[] parameters)
        {
            try
            {
                using (SqlConnection conn = new SqlConnection(connectionString))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(statement, conn))
                    {
                        if (parameters != null)
                        {
                            cmd.Parameters.AddRange(parameters);
                        }
                        return cmd.ExecuteNonQuery();
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error executing non-query: {ex.Message}");
                throw;
            }
            finally
            {
                // Additional cleanup if needed
            }
        }

        /// <summary>
        /// Execute a query and return the first column of the first row
        /// </summary>
        /// <param name="connectionString">SQL Server connection string</param>
        /// <param name="query">SQL query to execute</param>
        /// <param name="params">Parameters for the query</param>
        /// <returns>The value of the first column of the first row, or null</returns>
        public static object ExecuteScalar(string connectionString, string query, params SqlParameter[] parameters)
        {
            try
            {
                using (SqlConnection conn = new SqlConnection(connectionString))
                {
                    conn.Open();
                    using (SqlCommand cmd = new SqlCommand(query, conn))
                    {
                        if (parameters != null)
                        {
                            cmd.Parameters.AddRange(parameters);
                        }
                        return cmd.ExecuteScalar();
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error executing scalar: {ex.Message}");
                throw;
            }
            finally
            {
                // Additional cleanup if needed
            }
        }

        /// <summary>
        /// Get a list of all tables in the database
        /// </summary>
        /// <param name="connectionString">SQL Server connection string</param>
        /// <returns>List of table names</returns>
        public static List<string> GetTableNames(string connectionString)
        {
            List<string> tableNames = new List<string>();
            string query = "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE'";
            DataTable dt = QueryDatabase(connectionString, query);
            foreach (DataRow row in dt.Rows)
            {
                tableNames.Add(row["TABLE_NAME"].ToString());
            }
            return tableNames;
        }

        /// <summary>
        /// Check if a table exists in the database
        /// </summary>
        /// <param name="connectionString">SQL Server connection string</param>
        /// <param name="tableName">Name of the table to check</param>
        /// <returns>True if table exists, False otherwise</returns>
        public static bool TableExists(string connectionString, string tableName)
        {
            string query = "SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = @tableName AND TABLE_TYPE = 'BASE TABLE'";
            var param = new SqlParameter("@tableName", tableName);
            object result = ExecuteScalar(connectionString, query, param);
            return result != null;
        }

        /// <summary>
        /// Get the row count of a table or query
        /// </summary>
        /// <param name="connectionString">SQL Server connection string</param>
        /// <param name="tableOrQuery">Name of the table or a SQL SELECT query</param>
        /// <returns>Number of rows</returns>
        public static long GetRowCount(string connectionString, string tableOrQuery)
        {
            string query;
            if (tableOrQuery.Trim().ToUpper().StartsWith("SELECT"))
            {
                query = $"SELECT COUNT(1) FROM ({tableOrQuery}) AS subquery";
            }
            else
            {
                query = $"SELECT COUNT(1) FROM {tableOrQuery}";
            }
            
            object result = ExecuteScalar(connectionString, query);
            return Convert.ToInt64(result);
        }
    }
}
